_posts/2020-06-10-connect-kafka-spring-kafka.md

```diff
@@ -0,0 +1,575 @@
+    kafkaTemplate.setDefaultTopic("reflectoring-user");
+    return kafkaTemplate;
+  }
+  ...
+}
+```
+
+We can inject `MessageConverter` into our `KafkaTemplate` instance directly and the configured message converter must be compatible with the configured Kafka Serializer.
+
+In our example, we are using `StringJsonMessageConverter` with `StringSerializer`. Other compatible converters with their serializer are:
+
+- `StringJsonMessageConverter` with `StringSerializer`
+- `BytesJsonMessageConverter` with `BytesSerializer`
+- `ByteArrayJsonMessageConverter` with `ByteArraySerializer`
+
+Now that we have configured our produer to use `StringJsonMessageConverter`. We have to do the same for consumer as well:
+
+```java
+@Configuration
+class KafkaConsumerConfig {
+  @Bean
+  public ConcurrentKafkaListenerContainerFactory<String, String> kafkaJsonListenerContainerFactory() {
+  ...
+  factory.setMessageConverter(new StringJsonMessageConverter());
+  return factory;
+  }
+}
+```
+
+In the above code, we are creating a new `kafkaJsonListenerContainerFactory` with `StringJsonMessageConverter` set as our `MessageConverter`. Now let's create a listener to use the `kafkaJsonListenerContainerFactory` container factory:
+
+```java
+@Component
+class KafkaListenersExample {
+  ...
+
+  @KafkaListener(topics = "reflectoring-user", groupId="reflectoring-user-mc", containerFactory="kafkaJsonListenerContainerFactory")
+  void listenerWithMessageConverter(User user) {
+    LOG.info("MessageConverterUserListener [{}]", user);
+  }
+}
+```
+
+We can inject this created listener factory directly to our `@KafkaListener` with `containerFactory="kafkaJsonListenerContainerFactory"` as shown in the above example.
+
+Using `MessageConverter`, we can create any number of listeners for different types of records, without the need for multiple configurations for each type. When we use a `@KafkaListener`, the parameter type is provided to the method that will assist with the conversion (In our case the `User` type).
+
+Let us now send a record to the listener created above:
+
+```java
+@Component
+class KafkaSenderWithMessageConverter {
+
+  @Autowired
+  private KafkaTemplate<String,Message<?>> kafkaTemplate;
+
+  void sendMessageWithConverter(Message<?> user) {
+    kafkaTemplate.send(user);
+  }
+}
+```
+
+In the above example, we have created a `KafkaTemplate` which can be used to send any type of Java object. We just have to wrap the object to be sent inside Spring's `GenericMessage<>` before sending it as shown below.
+
+```
+messageConverterSender.sendMessageWithConverter(new GenericMessage<>(new User("Lucarioo")));
+```
+
+## Conclusion
+
+In this article, we covered how we can leverage the Spring support for Kafka. Build Kafka based messaging with code-examples that can help to get started quickly.
+
+{% include github-project.html url="https://github.com/thombergs/code-examples/tree/master/spring-boot/spring-events" %}
```

Comment:
I didn't mean to include the template again, but instead to include a simple link :). Did that now.

Reply:
Sorry, miss understood. 🤣